aodncore/pipeline/handlerbase.py

```diff
@@ -1022,6 +1022,18 @@ def postprocess(self):  # pragma: no cover
     # "public" methods
     #
 
+    def download_remotepipelinefilecollection(self, remotepipelinefilecollection, local_path=None):
```

Comment:
@mhidas - yeap I agree that it should also allow using the ```RemotePipelineFile``` object as arg.

However, when I mentioned, I was more thinking that ```RemotePipelineFile``` should also have a ```download``` method. This way, we could avoid initing/wrapping the thing into another class to just download a singleton file. 

This could affect the design of the Collection class though, and I understand if it's not quickly implemented.

Reply:
You can't download a file without knowing where to download it *from*. Having a remote file being downloadable in a self contained way means each file would need an independent link to a storage broker which is intentionally not implemented in this way. 

The relationship is one way, in that a broker knows how to download a RemotePipelineFileCollection from itself, and not the other way around. Otherwise every single class ends up doing absolutely everything, instead of being focused on one job. This is why the hierarchy is intentionally "one way" in terms of where the methods are.

Doing otherwise encourages child handlers to mess around in unnecessarily low level ways (such as trying to interact with storage directly) that are most likely inconsistent with other handlers and we get back to the scenario where handlers all go off and do their own similar things in different ways. Child handlers should deal with high level concepts through the library, and the library should deal with low level concepts such as storage and harvester implementation etc. Separation of concerns, abstraction, encapsulation, all that good stuff.

- A File represents a file and the state of that file.
- A FileCollection contains Files and understands how to filter and manipulate them en masse.
- A Broker can take a collection and attempt to download so that the collection object needs zero knowledge about the storage layer (since that is not it's purpose)
- The Handler in this case can simply pivot the parameters from being "tell this collection a broker it can use to download itself" (which requires knowing about brokers, and actually having one to use which is too low level) to "tell the handler to download this collection" (which requires zero low level knowledge)

So in a sense, if the child handler needs to upload or download, it must do this through the designated public library methods, otherwise it's breaking the intended encapsulation, and thinking in terms that are too "low level". Child handlers should be declarative in that sense. They should be defining what needs to be done, not *how* it is done.

Reply:
I'm talking about the ```download``` method in Collections being implemented in RemoteFile. After peeking through the code, I realize it should be simple.

My question was raised from the following observation:  ```RemoteFile``` can delete themselves, checksum themselves, so why they can't download themselves!? 

Then my thinking was: OK, no self.download method - let pick up the guy who can do it/define the location(broker). 

However, after the comment above, I checked the Collection method and the "download" method does exactly that - uses the broker to download. 

>The relationship is one way, in that a broker knows how to download a RemotePipelineFileCollection from itself, and not the other way around.

We are allowing both atm. ```RemotePipelineFileCollection.download``` allows doing exactly the opposite - in a semantic sense. You say "Hey RemotePipelineFileCollection, please download yourself with this broker at this path", but what happens is "Hey broker, use me and put the result on this path".

My whole point is, why not allow the same in RemoteFile!? Why do I need to wrap RemoteFile in another class, within a singleton IndexedSet, to just call a provided/required external object with a path argument? Just allow the same in RemoteFile then. 

>A Broker can take a collection and attempt to download so that the collection object needs zero knowledge about the storage layer (since that is not it's purpose)

Well, the collection group knows about who can help her being downloaded, so why the individual is ignorant?

Regards,
Nominated lawyer of RemoteFile rights

Reply:
Haha, it's been a while since thinking in the headspace. As usual, there is sometimes a discrepancy between my internal mental model and the current reality.

Now that I look at it, I didn't follow my own vision and implemented it wrongly and invalidated everything I just said ;)

> My question was raised from the following observation: RemoteFile can delete themselves, checksum themselves, so why they can't download themselves!?

The original intention for Files and Collections was that they can't *do* those things to themselves, they simply hold state information of their current status (e.g. has this been published) and intended state (e.g. is this a candidate for publishing). Some things crept in like checksumming itself and a remote file managing it's own locally cached file etc. but there is a real danger in this scope creeping out to all kinds of things. If anything this scope should be reduced back to being simpler types rather than expanded.

I actually think it would be better to get rid of the download method on the collection now that this is in place, because storage brokers really aren't something a collection type should have anything to do with.

Reply:
>Haha, it's been a while since thinking in the headspace. As usual, there is sometimes a discrepancy between my internal mental model and the current reality.
>Now that I look at it, I didn't follow my own vision and implemented it wrongly and invalidated everything I just said ;)
>The original intention for Files and Collections was that they can't do those things to themselves, they simply hold state information of their current status (e.g. has this been published) and intended state (e.g. is this a candidate for publishing). Some things crept in like checksumming itself and a remote file managing it's own locally cached file etc. but there is a real danger in this scope creeping out to all kinds of things. If anything this scope should be reduced back to being simpler types rather than expanded.

Maybe you don't need to prune but embrace the living objects (teenagers!?) they became and led them to adulthood!? IMO,  I don't think self.check_summing or self.download are jumping on the fence of their objectives, otherwise, they would be dumb down to just boring storage entities.  Maybe it's the right thing to do, but they will complain about losing privileges!

Reply:
> they would be dumb down to just boring storage entities

Boring sounds good to me... keep the exciting code out in the modules that deal with the data generation/transformation/aggregation, but by the time it's gets to the workflow it should be as generic, consistent and stable as possible!